sklearn/preprocessing/_encoders.py

```diff
@@ -123,6 +123,13 @@ def _fit(
                     )
                     raise ValueError(msg)
 
+                if len(cats) != len(set(cats)):
```

Comment:
Since we deal with a NumPy array, let's use numpy to solve this issue.
```suggestion
                _, n_unique_categories = np.unique(cats, return_counts=True)
                if cats.size != n_unique_categories:
```

Reply:
I think this causes the error. How about `if cats.size != np.unique(cats).size:`

Reply:
Yep `n_unique_categories = array([1, 1, 1, 1, 1])` is older NumPy version. I am not really sure why. The change that you propose is fine with me.

Reply:
Unfortunately, this still causes error due to `None`, see the fails.
For example, `np.unique(np.array([None, 'a', 'z'], dtype=object))` will raise an error.

How about `if cats.size != len(set(cats)):`?

Reply:
OK so let's revert to the set then. However, we need a test where we specify several time `nan` in the category as well to check this corner case.

Reply:
I am going to use `_unique` function instead, see the code below.

```python
import numpy as np
from sklearn.utils._encode import _unique

print(set(np.array(['a', None, None])))            # {'a', None}
print(set(np.array(['a', np.nan, np.nan])))        # {'a', 'nan'}
print(set(np.array([1., np.nan, np.nan])))         # {nan, 1.0, nan}

print(_unique(np.array(['a', None, None])))        # ['a' None]
print(_unique(np.array(['a', np.nan, np.nan])))    # ['a' 'nan']
print(_unique(np.array([1., np.nan, np.nan])))     # [ 1. nan]
```

For several `nan` in the category, I don't think it's necessary. Since now we assume `nan` must at the last, and PR #27309 will resolve this.

BTW, which way do you think is better? First check if `nan` is at the last or first check if category contain duplicated values?

Reply:
Hi @thomasjpfan, would you like take a look at this PR and also #27309 ?

Reply:
> BTW, which way do you think is better? First check if nan is at the last or first check if category contain duplicated values?

Maybe the check for `nan` first since it is less expensive.

Reply:
Sure, I will update code after #27309 is merged.